Documentation for local interconnected network setup (#3346)
hoprnet/hoprnet: SETUP_LOCAL_CLUSTER.md

# Introduction

For both HOPR app and core protocol developers, being able to spin a fully interconnected HOPR network is critical.
Without doing so, app developers will only be able to fake interactions with a HOPR node by mocking their API[^1], and
core protocol developers will not be able to replicate critical functionality over messaging such as ticket redemption
and packet relaying.

Although the HOPR Association provides a series of [nodes](https://status.hoprnet.org/) anyone could talk in production[^2],
this setup is not ideal for development as doing so would incur on costs for executing basic operations. Staging networks
such as görli are too slow to provide meaningful feedback, so the ideal setup for properly interacting with the HOPR
ecosystem toolset is running a local HOPR network, either in your workstation or via a local environment like
[Gitpod](https://gitpod.io). This way contributors are not constrained by internet connection or blockchain RPC providers.

# Tooling

> As of time of writing, the best way to setup a local HOPR network is by following these steps[^3]:

Author: ```suggestion
As of time of writing, the best way to setup a local HOPR network is by following these steps[^3]. Please bear in mind that these steps
had been tested only on `macOS` and `Linux` devices, for Windows please use our cloud setup instead.
```

   and unzip it in your local folder (roughly `~30 Mbs` at the time of writing). For the next tasks, we will assume you are within that folder.

```
wget https://github.com/hoprnet/hoprnet/archive/refs/heads/master.zip
unzip master.zip
cd hoprnet-master
```

2. **Install the dependencies of the project and build it**: Make sure you have `nodejs@16` (we suggest installing it via [nvm](https://github.com/nvm-sh/nvm), ie `nvm install lts/gallium`), and `yarn` (included in `nodejs@16` by running `corepack enable`).
   to install and build the required packages and project modules. Please bear in mind that this process will take at least 5-10 minutes depending on your computer.

```
yarn && yarn build
```

3. **Run the one-line setup script**: Proceed to run the following script. If you are planning to run [MyneChat](http://app.myne.chat/)
   alongside, then make sure to pass the `-m` flag with your MyneChat instance URL. Please wait while this script creates
   the local blockchain network and deploys the project contracts. In average, the process can take between 2-6 minutes,

>    depending on your computer capacity. **Important**, make sure to have both `curl` and `jq` installed in your computer before running the script, as both are used.

Author: ```suggestion
   depending on your computer capacity. **Important**, make sure to have both `curl` and `jq` installed in your computer
   before running the script, as both are used. Please be aware you also need a version of `bash` of `5.x` or superior,
   which in most macOS devices require an upgrade, easiest being via `brew bash`.
```

core protocol developers will not be able to replicate critical functionality over messaging such as ticket redemption
and packet relaying.

Although the HOPR Association provides a series of [nodes](https://status.hoprnet.org/) anyone could talk in production[^2],
this setup is not ideal for development as doing so would incur on costs for executing basic operations. Staging networks
such as görli are too slow to provide meaningful feedback, so the ideal setup for properly interacting with the HOPR
ecosystem toolset is running a local HOPR network, either in your workstation or via a local environment like
[Gitpod](https://gitpod.io). This way contributors are not constrained by internet connection or blockchain RPC providers.

# Tooling

As of time of writing, the best way to setup a local HOPR network is by following these steps[^3]:

1. **Download the latest version of the repository**: Download a local version of our [GitHub repository](https://github.com/hoprnet/hoprnet)[^4]
   and unzip it in your local folder. For the next tasks, we will assume you are within that folder.

```

> wget https://github.com/hoprnet/hoprnet/archive/refs/heads/master.zip

Author: Use a specific version instead of `master`, see hoprnet/hopr-devrel#167

Author: Since we'll tackle this in https://github.com/hoprnet/hopr-devrel/issues/167, we can resolve this one there.

unzip master.zip
cd hoprnet-master
```

2. **Install the dependencies of the project and build it**: Make sure you are using `nodejs@16`, and proceed to use `yarn`
   to install the required packages and project modules. Soon afterward, build the project to use it in the next step.

```
yarn

> yarn build

Reviewer: It's probably worth adding sth like `apt install build-essentials` since `yarn` builds some C / C++ code under the hood

```

3. **Run the one-line setup script**: Proceed to run the following script. If you are planning to run [MyneChat](http://app.myne.chat/)
   alongside, then make sure to pass the `-m` flag with your MyneChat instance URL. Please wait while this script creates
   the local blockchain network and deploys the project contracts. In average, the process can take between 2-6 minutes,

>    depending on your computer capacity.

Author: ```suggestion
   depending on your computer capacity. **Important**, make sure to have both `curl` and `jq` installed in your computer before running the script, as both are used.
```

# Introduction

> For both HOPR app and core protocol developers, being able to spin a fully interconnected HOPR network is critical.

Reviewer: Who is the recipient of this text?

Author: `app` and `core protocol` developers, I believe.

Reviewer: Might sound strange, but isn't this a little bit too direct? It feels like someone explaining me how to do my job

Author: Fair take, I can soften it up, one sec.

Author: Improved this in 90d76bf, please feel free to provide some feedback there.

# Tooling

As of time of writing, the best way to setup a local HOPR network is by following these steps[^3]. Please bear in mind that these steps
had been tested only on `macOS` and `Linux` devices, for Windows please use our cloud setup instead.

1. **Download the latest version of the repository**: Download a local version of our [GitHub repository monorepo](https://github.com/hoprnet/hoprnet)[^4]
   and unzip it in your local folder (roughly `~30 Mbs` at the time of writing). For the next tasks, we will assume you are within that folder.

```
wget https://github.com/hoprnet/hoprnet/archive/refs/heads/master.zip
unzip master.zip
cd hoprnet-master
```

2. **Install the dependencies of the project and build it**: Make sure you have `nodejs@16` (we suggest installing it via [nvm](https://github.com/nvm-sh/nvm), ie `nvm install lts/gallium`), and `yarn` (included in `nodejs@16` by running `corepack enable`).
   to install and build the required packages and project modules. Please bear in mind that this process will take at least 5-10 minutes depending on your computer.

```

> yarn && yarn build

Reviewer: `yarn build` requires on Ubuntu the package `build-essentials` and on MacOS the XCode CLI Tools

Author: I spun up a brand new VM with `Ubuntu 20.04` and didn't install it. It might have been pre-installed as part of the installation process. Either way, I think it's worth adding just in case, let me add a footnote there.

# Introduction

> For both HOPR app and core protocol developers, being able to spin a fully interconnected HOPR network is critical.

Author: ```suggestion
The best way to learn and develop the HOPR protocol is by interacting with a HOPR node connected with the HOPR network. Since the HOPR network requires real money to interact, the alternative involves spinning a HOPR cluster locally or within a cloud setup to mimic the same setup.
```

> Without doing so, app developers will only be able to fake interactions with a HOPR node by mocking their API[^1], and

Author: ```suggestion
Without a HOPR cluster, app developers will only be able to fake interactions with a HOPR node by mocking their API[^1], and
```

> core protocol developers will not be able to replicate critical functionality over messaging such as ticket redemption

Author: ```suggestion
core protocol developers can't replicate critical functionality over messaging such as ticket redemption
```

and packet relaying.

> Although the HOPR Association provides a series of [nodes](https://status.hoprnet.org/) anyone could talk in production[^2],

Author: ```suggestion
Although the HOPR Association provides a [production HOPR cluster](https://status.hoprnet.org/) anyone could talk in production[^2],
```

this setup is not ideal for development as doing so would incur on costs for executing basic operations. Staging networks
such as görli are too slow to provide meaningful feedback, so the ideal setup for properly interacting with the HOPR

> ecosystem toolset is running a local HOPR network, either in your workstation or via a local environment like

Author: ```suggestion
ecosystem toolset is running a local HOPR cluster, either in your workstation or via a cloud environment like
```

[Gitpod](https://gitpod.io). This way contributors are not constrained by internet connection or blockchain RPC providers.

# Tooling

As of the time of writing, the best way to set up a local HOPR cluster is by following these steps[^3]. Please bear in mind that these steps
had been tested only on `macOS` and `Linux` devices, for Windows please use our cloud setup instead.

1. **Download the latest version of the repository**: Download a local version of our [GitHub repository monorepo](https://github.com/hoprnet/hoprnet)[^4]
   and unzip it in your local folder (roughly `~30 Mbs` at the time of writing). For the next tasks, we will assume you are within that folder.

```

> wget https://github.com/hoprnet/hoprnet/archive/refs/heads/master.zip

Reviewer 2: why not use git? as we are targeting developers

Author: Cloning with `git` takes way more space since it also downloads all the git history.

> wget https://github.com/hoprnet/hoprnet/archive/refs/heads/master.zip

Reviewer 2: why not use git? as we are targeting developers

   depending on your computer capacity. **Important**, make sure to have both `curl` and `jq` installed in your computer
   before running the script, as both are used. Please be aware you also need a version of `bash` of `5.x` or superior,
   which in most macOS devices require an upgrade, the easiest being via `brew bash`.

```
./scripts/setup-local-cluster.sh -m "http://app.myne.chat" -i scripts/topologies/full_interconnected_cluster.sh
```

Afterwards, a set off accounts with their respective HTTP REST API, HOPR Admin, and WebSocket interface will be displayed
in your screen. If this is your first time using HOPR, I suggest navigating to the `HOPR Admin` URL to get familiar with
the basic commands. Afterwards, it might also make sense to check the API v2 Swagger URL.

[^1]:
    The demo application [MyneChat](https://github.com/hoprnet/myne-chat) uses a
    [mock server](https://github.com/hoprnet/myne-chat/blob/cf6501b2ffa24502834f567ab575630e302e3d34/mocks/index.js#L47-L79)
    to simplify it’s development workflow. Nevertheless, to fully experience the extend of its features, it relies on a

>     HOPR network, either a local or a public one.

Author: ```suggestion
    HOPR cluster, either a local or a public one.
```
